src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -214,76 +205,82 @@ void Transceiver::stop_sem_cb_(uv_async_t* handle) {
+}
+
 void Transceiver::stop_() {
-    for (core::SharedPtr<UDPReceiver> rp = receivers_.front(); rp;
-         rp = receivers_.nextof(*rp)) {
-        rp->stop();
+    core::Mutex::Lock lock(mutex_);
+
+    if (stopped_) {
+        return;
     }
 
-    for (core::SharedPtr<UDPSender> sp = senders_.front(); sp;
-         sp = senders_.nextof(*sp)) {
-        sp->stop();
+    if (int err = uv_async_send(&stop_sem_)) {
+        roc_panic("transceiver: uv_async_send(): [%s] %s", uv_err_name(err),
+                  uv_strerror(err));
     }
+
+    wait_stopped_();
+    stopped_ = true;
 }
 
-void Transceiver::close_() {
-    if (task_sem_initialized_) {
-        uv_close((uv_handle_t*)&task_sem_, NULL);
-        task_sem_initialized_ = false;
+void Transceiver::wait_stopped_() {
+    while (ports_.size() || task_sem_initialized_ || stop_sem_initialized_) {
+        cond_.wait();
     }
+}
 
-    if (stop_sem_initialized_) {
-        uv_close((uv_handle_t*)&stop_sem_, NULL);
-        stop_sem_initialized_ = false;
+void Transceiver::async_close_ports_() {
+    core::SharedPtr<BasicPort> curr = ports_.front();
+    while (curr) {
+        core::SharedPtr<BasicPort> next = ports_.nextof(*curr);
+        curr->async_close();
+        curr = next;
     }
 }
 
-void Transceiver::remove_all_ports_() {
-    core::SharedPtr<UDPReceiver> rp = receivers_.front();
-    while (rp) {
-        core::SharedPtr<UDPReceiver> rp_next = receivers_.nextof(*rp);
-        rp->remove(receivers_);
-        rp = rp_next;
-        num_ports_--;
+void Transceiver::close_sems_() {
+    if (task_sem_initialized_) {
+        uv_close((uv_handle_t*)&task_sem_, close_sems_cb_);
     }
 
-    core::SharedPtr<UDPSender> sp = senders_.front();
-    while (sp) {
-        core::SharedPtr<UDPSender> sp_next = senders_.nextof(*sp);
-        sp->remove(senders_);
-        sp = sp_next;
-        num_ports_--;
+    if (stop_sem_initialized_) {
+        uv_close((uv_handle_t*)&stop_sem_, close_sems_cb_);
     }
 }
 
 void Transceiver::run_task_(Task& task) {
     core::Mutex::Lock lock(mutex_);
 
-    const bool running = joinable();
+    roc_panic_if(stopped_);
 
-    if (!running || stopped_) {
-        // If a stop was scheduled, ensure event loop thread have finished.
-        if (running) {
-            mutex_.unlock();
-            join();
-            mutex_.lock();
-        }
-
-        // There is no event loop thread, execute task in-place.
-        task.execute(*this);
-    } else {
-        // Schedule task on event loop thread.
-        tasks_.push_back(task);
+    tasks_.push_back(task);
 
-        if (int err = uv_async_send(&task_sem_)) {
-            roc_panic("transceiver: uv_async_send(): [%s] %s", uv_err_name(err),
-                      uv_strerror(err));
-        }
+    if (int err = uv_async_send(&task_sem_)) {
+        roc_panic("transceiver: uv_async_send(): [%s] %s", uv_err_name(err),
+                  uv_strerror(err));
     }
 
     while (!task.done) {
```

Comment:
I see no place when `done` is set. How is it working? Why tests are not failing?